internal/repository/master_index.go

```diff
@@ -65,11 +66,47 @@ func (mi *MasterIndex) ListPack(id restic.ID) (list []restic.PackedBlob) {
 	return nil
 }
 
+// CheckSetKnown returns wheter the blob is already known to the index
+// A known blob is a blob that is either in the index or has been marked
+// as known previously.
+//
+// If the given blob not known, CheckSetKnown returns false
+// and marks this blob as known.
+//
+// A marked blob is unmarked once it is stored in the index.
+func (mi *MasterIndex) CheckSetKnown(id restic.ID, tpe restic.BlobType) bool {
+
+	if mi.Has(id, tpe) {
+		return true
+	}
+
+	// Seems it is not known; lock to add blob to known
+	mi.idxMutex.Lock()
```

Comment:
I realized that there is that issue. IMO the same issue is already present in upstream master...
I however think this can be easily changed by always using a write lock...